src/compute-client/src/controller.rs

```diff
@@ -163,9 +169,357 @@ impl From<anyhow::Error> for ComputeError {
+    /// Return all persisted introspection sources contained
+    pub fn get_sources(&self) -> Vec<(LogVariant, GlobalId)> {
+        match self {
+            ConcreteComputeInstanceReplicaLogging::Default => vec![],
+            ConcreteComputeInstanceReplicaLogging::ConcreteViews(logs, _) => logs.clone(),
+        }
+    }
+
+    /// Return all persisted introspection views contained
+    pub fn get_views(&self) -> Vec<(LogView, GlobalId)> {
+        match self {
+            ConcreteComputeInstanceReplicaLogging::Default => vec![],
+            ConcreteComputeInstanceReplicaLogging::ConcreteViews(_, views) => views.clone(),
+        }
+    }
+
+    /// Return all ids of the persisted introspection views contained
+    pub fn get_view_ids(&self) -> Vec<GlobalId> {
+        self.get_views().into_iter().map(|(_, id)| id).collect()
+    }
+
+    /// Return all ids of the persisted introspection sources contained
+    pub fn get_source_ids(&self) -> Vec<GlobalId> {
+        self.get_sources().into_iter().map(|(_, id)| id).collect()
+    }
+
+    /// Return all ids of the persisted introspection sources and logs contained
+    pub fn get_source_and_view_ids(&self) -> Vec<GlobalId> {
+        self.get_source_ids()
+            .into_iter()
+            .chain(self.get_view_ids().into_iter())
+            .collect()
+    }
+}
+
+/// A controller for the compute layer.
+pub struct ComputeController<T> {
+    instances: BTreeMap<ComputeInstanceId, Instance<T>>,
+    build_info: &'static BuildInfo,
+    orchestrator: ComputeOrchestrator,
+    /// Set to `true` once `initialization_complete` has been called.
+    initialized: bool,
+}
 
 impl<T> ComputeController<T> {
+    /// Construct a new [`ComputeController`].
+    pub fn new(
+        build_info: &'static BuildInfo,
+        orchestrator: Arc<dyn NamespacedOrchestrator>,
+        computed_image: String,
+    ) -> Self {
+        Self {
+            instances: BTreeMap::new(),
+            build_info,
+            orchestrator: ComputeOrchestrator::new(orchestrator, computed_image),
+            initialized: false,
+        }
+    }
+
+    /// Return a handle to the indicated compute instance, if it exists.
+    pub fn instance(&self, id: ComputeInstanceId) -> Option<&Instance<T>> {
+        self.instances.get(&id)
+    }
+
+    /// Acquire an [`ActiveComputeController`] by supplying a storage connection.
+    pub fn activate<'a>(
+        &'a mut self,
+        storage: &'a mut dyn StorageController<Timestamp = T>,
+    ) -> ActiveComputeController<'a, T> {
+        ActiveComputeController {
+            compute: self,
+            storage,
+        }
+    }
+}
+
+impl<T> ComputeController<T>
+where
+    T: Timestamp + Lattice,
+    ComputeGrpcClient: ComputeClient<T>,
+{
+    /// Create a compute instance.
+    pub fn create_instance(&mut self, instance: ComputeInstanceId, logging: Option<LoggingConfig>) {
+        // Insert a new compute instance controller.
```

Comment:
Not sure this is new, but shouldnt we check that instance is not yet ins `self.instances`?

Reply:
Yeah, this was moved from src/controller/src/lib.rs. But the answer do your question is probably "yes, we should"!